core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchAppRequestClient.java

```diff
@@ -0,0 +1,180 @@
+      final String baseUrl,
+      final Transport transport,
+      final Long defaultRequestTimeout
+  ) {
+    super(baseUrl, transport, defaultRequestTimeout);
+    this.clientAppId = clientAppId;
+  }
+
+  /**
+   * Performs a request against a Stitch app server determined by the deployment model
+   * of the underlying app. Throws a Stitch specific exception if the request fails.
+   *
+   * @param stitchReq the request to perform.
+   * @return a {@link Response} to the request.
+   */
+  @Override
+  public Response doRequest(final StitchRequest stitchReq) {
+    if (appMetadata == null) {
+      init(clientAppId);
+    }
+    // init throws if no location is available, so this should be safe
+
+    final Response response;
+    try {
+      response = transport.roundTrip(buildRequest(stitchReq, appMetadata.hostname));
+    } catch (Exception e) {
+      throw new StitchRequestException(e, StitchRequestErrorCode.TRANSPORT_ERROR);
+    }
+
+    inspectResponse(response);
+
+    return response;
+  }
+
+  /**
+   * Performs a streaming request against a Stitch app server determined by the deployment model
+   * of the underlying app. Throws a Stitch specific exception if the request fails.
+   *
+   * @param stitchReq the request to perform.
+   * @return a {@link EventStream} representing the result of the request.
+   */
+  @Override
+  public EventStream doStreamRequest(final StitchRequest stitchReq) {
+    if (appMetadata == null) {
+      init(clientAppId);
+    }
+    // init throws if no location is available, so this should be safe
+
+    try {
+      return transport.stream(buildRequest(stitchReq, appMetadata.hostname));
+    } catch (Exception e) {
+      throw new StitchRequestException(e, StitchRequestErrorCode.TRANSPORT_ERROR);
+    }
+  }
+
+  private synchronized void init(final String clientAppId) {
+    // double-check, in case two threads hit this point simultaneously
+    if (appMetadata != null) {
+      return;
+    }
+
+    final StitchAppRoutes routes = new StitchAppRoutes(clientAppId);
+
+    final StitchRequest bootstrapStitchRequest = new StitchRequest.Builder()
+        .withMethod(Method.GET)
+        .withPath(routes.getServiceRoutes().getLocationRoute())
+        .build();
+    final Request bootstrapRequest = buildRequest(bootstrapStitchRequest, baseUrl);
+
+    final Response response;
+    try {
+      response = transport.roundTrip(bootstrapRequest);
+    } catch (Exception e) {
+      throw new StitchRequestException(e, StitchRequestErrorCode.TRANSPORT_ERROR);
+    }
+
+    inspectResponse(response);
+
+    final AppMetadata responseMetadata;
+    try {
+      responseMetadata = StitchObjectMapper.getInstance()
+          .readValue(IoUtils.readAllToString(response.getBody()), AppMetadata.class);
+
+    } catch (Exception e) {
+      throw new StitchRequestException(e, StitchRequestErrorCode.DECODING_ERROR);
+    }
+
+    if (responseMetadata != null && responseMetadata.hostname != null) {
```

Comment:
[nit] In addition to checking the hostname is not null, I'd check that it's also non empty.

Reply:
done